File: python/200岛屿数量.py
class Solution(object):
    def dfs(self, grid, r, c):
        #设置该位置为0
        grid[r][c] = 0
        m = len(grid)
        n = len(grid[0])
        #搜索四周
        for x, y in [(r-1, c), (r+1, c), (r, c-1), (r, c+1)]:
            if 0<= x < m and 0<= y < n and grid[x][y] == "1":
                self.dfs(grid, x, y) #递归

    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        m = len(grid)
        if m == 0:
            return 0
        n = len(grid[0])
        count = 0
        for r in range(m):
            for c in range(n):
                if grid[r][c] == "1":
                    count += 1
                    self.dfs(grid, r, c)
        return count

File: java/200岛屿数量.java
class Solution {
    public void dfs(char[][] grid, int r, int c){
        int m = grid.length;
        int n = grid[0].length;
        if(r < 0 || c < 0 || r >= m || c >= n || grid[r][c] == '0'){
            return;
        }
        grid[r][c] = '0';
        dfs(grid, r-1, c);
        dfs(grid, r+1, c);
        dfs(grid, r, c-1);
        dfs(grid, r, c+1);

    }

    public int numIslands(char[][] grid) {
        int m = grid.length;
        if(m == 0){
            return 0;
        }
        int n = grid[0].length;
        int count = 0;
        for(int r = 0; r < m; ++r){
            for(int c = 0; c < n; ++c){
                if(grid[r][c] == '1'){
                    count += 1;
                    dfs(grid, r, c);
                }
            }
        }
        return count;
    }
}